tlspuffin/harness/openssl/src/put.c

```diff
@@ -383,18 +357,75 @@ static AGENT make_agent(SSL_CTX *ssl_ctx, const TLS_AGENT_DESCRIPTOR *descriptor
     return agent;
 }
 
-static RESULT_CODE result_code(AGENT agent, int retcode)
+static RESULT get_result(AGENT agent, int retcode, bool allow_would_block)
 {
     int ssl_ecode = SSL_get_error(agent->ssl, retcode);
+
+    char *reason = get_error_reason();
+    char *error_type;
+    RESULT_CODE res = RESULT_OK;
+
     switch (ssl_ecode)
     {
     case SSL_ERROR_NONE:
-        return RESULT_OK;
-
+        error_type = strdup("no error");
+        break;
+    case SSL_ERROR_ZERO_RETURN:
+        error_type = strdup("SSL_ERROR_ZERO_RETURN");
+        break;
+    case SSL_ERROR_WANT_CONNECT:
+        error_type = strdup("SSL_ERROR_WANT_CONNECT");
+        break;
+    case SSL_ERROR_WANT_ACCEPT:
+        error_type = strdup("SSL_ERROR_WANT_ACCEPT");
+        break;
+    case SSL_ERROR_WANT_X509_LOOKUP:
+        error_type = strdup("SSL_ERROR_WANT_X509_LOOKUP");
+        break;
+    case SSL_ERROR_WANT_ASYNC:
+        error_type = strdup("SSL_ERROR_WANT_ASYNC");
+        break;
+    case SSL_ERROR_WANT_ASYNC_JOB:
+        error_type = strdup("SSL_ERROR_WANT_ASYNC_JOB");
+        break;
+    case SSL_ERROR_WANT_CLIENT_HELLO_CB:
+        error_type = strdup("SSL_ERROR_WANT_CLIENT_HELLO_CB");
+        break;
+    case SSL_ERROR_SYSCALL:
+        error_type = strdup("SSL_ERROR_SYSCALL");
+        break;
+    case SSL_ERROR_SSL:
+        error_type = strdup("SSL_ERROR_SSL");
```

Comment:
This should be considered equivalent to a crash!
See from https://docs.openssl.org/1.1.1/man3/SSL_get_error/#return-values:
SSL_ERROR_SSL

A non-recoverable, fatal error in the SSL library occurred, usually a protocol error. The OpenSSL error queue contains more information on the error. If this error occurs then no further I/O operations should be performed on the connection and SSL_shutdown() must not be called.

Reply:
As discussed, an SSL error is not a crash, since it has been catched by OpenSSL : the server should abort the current connection but is not in a state of DoS. Furthermore, we want this to be non-blocking to see if we can still exchange messages with the PUT once it is in an error state to potentially find bugs.